File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/controls_for.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeControlsFor(statement: JSONObject, onError: (e: Exception) -> Unit) {
    try {
        Log.d("hello", "execute_controls_forEach")
        loopFlow = null
        val pointer = statement.getJSONObject("pointer")
        val start = executeStatement(pointer["from"].toString()).toString().toInt()
        val end = executeStatement(pointer["to"].toString()).toString().toInt()
        val step = executeStatement(pointer["step"].toString()).toString().toInt()

        val variableName = pointer["variable"].toString()
        val loopBlock = pointer["loopBlock"].toString().trim()

        for (i in start..end step step) {
            VARIABLE_STACK[variableName] = i

            if (loopBlock != "") execute(loopBlock)

            if (isTerminated) break
            if (loopFlow == "BREAK") {
                loopFlow = null
                break
            } else if (loopFlow == "CONTINUE") {
                loopFlow = null
                continue
            }

            if (isFunctionReturn != null) break
        }
    } catch (e: Exception) {
        onError(e)
    }
}

File: piper/extensions/chrome/src/engine/statements/controls_for.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        this.loopFlow = null
        let pointer = statement.pointer
        let start = await this.executeStatement(stringify(pointer.from))
        start = parseInt(start)
        let end = await this.executeStatement(stringify(pointer.to))
        end = parseInt(end)
        let step = await this.executeStatement(stringify(pointer.step))
        step = parseInt(step)

        let variableName = stringify(pointer.variable)
        let loopBlock = stringify(pointer.loopBlock).trim()

        for (let i = start; i <= end; i += step) {
            this.VARIABLE_STACK[variableName] = i

            if (loopBlock !== "") await this.execute(loopBlock)
            
            if (this.isTerminated) break
            if (this.loopFlow === "BREAK") {
                this.loopFlow = null
                break
            } else if (this.loopFlow === "CONTINUE") {
                this.loopFlow = null
                continue
            }

            if (this.isFunctionReturn !== null) break
        }
    } catch (e) {
        onError(e)
    }
}
